core/src/main/java/io/questdb/cairo/wal/seq/TableTransactionLog.java

```diff
@@ -0,0 +1,359 @@
+    private static final ThreadLocal<TableMetadataChangeLogImpl> tlStructChangeCursor = new ThreadLocal<>();
+    private final FilesFacade ff;
+    private final MemoryCMARW txnMem = Vm.getCMARWInstance();
+    private final MemoryCMARW txnMetaMem = Vm.getCMARWInstance();
+    private final MemoryCMARW txnMetaMemIndex = Vm.getCMARWInstance();
+    private final StringSink rootPath = new StringSink();
+    private long maxTxn;
+
+    TableTransactionLog(FilesFacade ff) {
+        this.ff = ff;
+    }
+
+    @Override
+    public void close() {
+        Misc.free(txnMem);
+        Misc.free(txnMetaMem);
+        Misc.free(txnMetaMemIndex);
+    }
+
+    @NotNull
+    static TableMetadataChangeLog getTableMetadataChangeLog() {
+        TableMetadataChangeLogImpl instance = tlStructChangeCursor.get();
+        if (instance == null) {
+            tlStructChangeCursor.set(instance = new TableMetadataChangeLogImpl());
+        }
+        return instance;
+    }
+
+    private static long openFileRO(final FilesFacade ff, final Path path, final String fileName) {
+        final int rootLen = path.length();
+        path.concat(fileName).$();
+        try {
+            return TableUtils.openRO(ff, path, LOG);
+        } finally {
+            path.trimTo(rootLen);
+        }
+    }
+
+    long addEntry(int walId, int segmentId, long segmentTxn) {
+        txnMem.putInt(walId);
+        txnMem.putInt(segmentId);
+        txnMem.putLong(segmentTxn);
+        txnMem.jumpTo(txnMem.getAppendOffset() + RECORD_RESERVED);
+
+        Unsafe.getUnsafe().storeFence();
+        txnMem.putLong(MAX_TXN_OFFSET, ++maxTxn);
+        // Transactions are 1 based here
+        return maxTxn;
+    }
+
+    long beginMetadataChangeEntry(long newStructureVersion, MemorySerializer serializer, Object instance) {
+        assert newStructureVersion == txnMetaMemIndex.getAppendOffset() / Long.BYTES;
+
+        txnMem.putInt(STRUCTURAL_CHANGE_WAL_ID);
+        txnMem.putLong(newStructureVersion);
+        txnMem.jumpTo(++maxTxn * RECORD_SIZE + HEADER_SIZE);
+
+        txnMetaMem.putInt(0);
+        long varMemBegin = txnMetaMem.getAppendOffset();
+        serializer.toSink(instance, txnMetaMem);
+        int len = (int) (txnMetaMem.getAppendOffset() - varMemBegin);
+        txnMetaMem.putInt(varMemBegin - Integer.BYTES, len);
+        txnMetaMemIndex.putLong(varMemBegin + len);
+
+        return varMemBegin + len;
+    }
+
+    long endMetadataChangeEntry(long newStructureVersion, long offset) {
+        Unsafe.getUnsafe().storeFence();
+        txnMem.putLong(MAX_TXN_OFFSET, maxTxn);
+        txnMem.putLong(MAX_STRUCTURE_VERSION_OFFSET, newStructureVersion);
+        txnMem.putLong(TXN_META_SIZE_OFFSET, offset);
+
+        // Transactions are 1 based here
+        return maxTxn;
+    }
+
+    TransactionLogCursor getCursor(long txnLo) {
+        final Path path = Path.PATH.get().of(rootPath);
+        final int rootLen = path.length();
+        path.concat(TXNLOG_FILE_NAME).$();
```

Comment:
We concat `TXNLOG_FILE_NAME` two times in this method: here and later in `new TransactionLogCursorImpl()`. This broke a test in my WAL + ILP branch.

Reply:
ok, fixed

Reply:
Thanks!